Block warning: add ellipsis menu and convert to classic block

johngodley opened the pull request:
## Description
Add the ability for a block warning message to show a list of additional actions that are hidden behind a dropdown ellipsis menu. This will be used to expand the options available to a user when an invalid block is detected.

The '_convert to blocks_' option is copied into this menu, along with a new '_convert to classic block_' option. This matches the mockup in #7604. More options will be added in future PRs

![another_post___ _wordpress_latest_ _wordpress](https://user-images.githubusercontent.com/1277682/43129377-2d237a76-8f2c-11e8-865a-74ce0525162b.jpg)

## How has this been tested?

Extra unit tests have been added to verify when the ellipsis menu appears.

`npm run test warning`

## Types of changes

A non-breaking new feature is added to an existing component. The `Warning` component is used by:

- `BlockInvalidWarning`
- `BlockCrashWarning`

To test `BlockInvalidWarning`:
- Create a paragraph
- Edit as HTML and remove the trailing `</p>` to trigger the block validation message
- Verify that the invalid block message is displayed with a new ellipsis menu:
![another_post___ _wordpress_latest_ _wordpress](https://user-images.githubusercontent.com/1277682/42573058-fbded00a-8512-11e8-9cf8-e09fa7ef20d4.jpg)
- Verify that reducing the page width causes the primary buttons to move onto the next line, but the ellipsis menu remains on the right:
![another_post___ _wordpress_latest_ _wordpress_and_gutenberg_ _update_block-warning-options__origin_update_block-warning-options__7169_commits_](https://user-images.githubusercontent.com/1277682/42573270-86a6348a-8513-11e8-945f-837c24ab975a.jpg)
- Verify that clicking 'convert to block' from the ellipsis menu performs the same action as the 'convert to block' button
- Verify that clicking 'convert to classic block' converts it to a classic block

To test `BlockCrashWarning`:
- From a developer console type `wp.blocks.registerBlockType( 'myplugin/mr-crashy', { title: 'Mr. Crashy', category: 'common', icon: 'warning', edit() { throw new Error(); }, save() { return ''; } } );`
- Add `Mr Crashy` block to a post
- Verify that the crash message is displayed without changes:
![another_post___ _wordpress_latest_ _wordpress](https://user-images.githubusercontent.com/1277682/42457439-6470a68c-838f-11e8-9cf8-3b6a4f49f695.jpg)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

jasmussen commented:
Very nice work, I pushed some polish:

![screen shot 2018-07-06 at 13 31 31](https://user-images.githubusercontent.com/1204802/42376869-0bb2e70a-8121-11e8-8b31-7266e601a976.png)

I also removed the styles for the contents of the navibable menu. Can I kindly ask you to take a look at [the more menu](https://github.com/WordPress/gutenberg/tree/master/edit-post/components/header/more-menu) and perhaps steal a bit from there? We have some components for menuy groups, menu labels and even menu items. If we reuse those, we get the visual styles, as well as their accessibility for free.

johngodley committed:
Block warning: add option for hidden actions

Adds an ellipsis dropdown menu so additional actions can be added outside of the primary buttons

Note this just adds the capability for additional actions, but doesn’t actually include any

johngodley committed:
Block warning: add convert to classic option

Makes use of the new ellipsis menu to present the sa,e convert to blocks option along with a convert to classic block

johngodley committed:
Tweak block warning dropdown menu position

The menu is slightly vertically off-centred when on full screen desktop. This moves it a bit while still looking ok on mobile

youknowriad commented on packages/editor/src/components/block-list/block-invalid-warning.js:
```diff
@@ -30,6 +30,10 @@ function BlockInvalidWarning( { convertToHTML, convertToBlocks } ) {
 					</Button>
 				),
 			] }
+			hiddenActions={ [
```
Could this be secondaryActions (to match the "primaryActions" prop)?

youknowriad commented on packages/editor/src/components/warning/index.js:
```diff
@@ -7,23 +7,53 @@ import classnames from 'classnames';
+			{ hiddenActions && (
+				<div className="editor-warning__hidden">
+					<Dropdown
+						className="edit-post-more-menu"
```
The `edit-post-*` className seems useless right? (they don't follow our guidelines neither)

johngodley replied:
It's the same class used in https://github.com/WordPress/gutenberg/blob/master/edit-post/components/header/more-menu/index.js, making the ellipsis vertical rather than horizontal. I can file a separate issue if that also needs changing to match the guidelines?

youknowriad replied:
I don't think we should rely on classnames styled in other components. We should aim to reuse components instead of classnames instead.

So two options:

 - I'm fine duplicating the styles (if not too much) and using the right classnames here `editor-warning__*`
 - Creating a generic component with these styles but in that case, this component should be meaningful. Maybe it could be a prop of `Dropdown` instead?

youknowriad reviewed (commented):
Looking good, just left some small notes.

johngodley committed:
Change hidden actions to secondary actions

Matches primary actions

johngodley committed:
Fix unit test to match secondaryActions change

johngodley committed:
Split warning dropdown style away from more dropdown

CSS is minimal, and can be refactored into a new component later

johngodley committed:
Update unit test to match new class name

johngodley commented:
@youknowriad thanks for looking. All good points.

For the purpose of this PR I've copied the appropriate styles to this component - it's pretty minimal.

While looking at the CSS I noticed the original component has what appears to be dead CSS as well as problems on mobile (that only affect the other component).

I'll investigate that separately and if it makes sense, refactor with this component while fixing any problems.

youknowriad reviewed (approved):
LGTM 👍 Nice work

aduth commented:
This shouldn't have been merged.

- It broke backwards-compatibility with the `Warning` `actions` prop by renaming it to `primaryActions`. This should have been deprecated, with support for `actions` kept for two versions, documented in `docs/deprecated.md` and noted in `packages/editor/CHANGELOG.md`
- Related to the previous point, I assume it broke both `ErrorBoundary` and the "validate multiple use" hook, both of which render `Warning` with the now-non-existant `actions` prop
   - https://github.com/WordPress/gutenberg/blob/6259b235ce0d1fa2fae597ebbbc176798bc67010/packages/editor/src/components/error-boundary/index.js#L52-L54
   - https://github.com/WordPress/gutenberg/blob/6259b235ce0d1fa2fae597ebbbc176798bc67010/edit-post/hooks/validate-multiple-use/index.js#L75-L77

youknowriad commented:
Good catch @aduth I missed that :(
